feat(lsp): Improve vim.lsp.util.apply_text_edits

hrsh7th opened the pull request:
This PR aims to solve the reversed range of TextEdit and including some of the refactor.

- Fix cursor position after TextEdit
    - This is the important thing because the `rust-analyzer` sends `additoinalTextEdits` that depending on this behavior
- Invoke nvim_buf_set_text one by one
    - So `extmarks` will be shifted correctly.

I will use it in nvim-cmp after this PR merged.

- [x] Add cursor position fixing

Fixes #15467

hrsh7th commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -255,45 +252,73 @@ function M.apply_text_edits(text_edits, bufnr)
-  end
+  -- Fix reversed range and indexing each text_edits
+  local index = 0
+  text_edits = vim.tbl_map(function(text_edit)
```
See https://github.com/microsoft/vscode-languageserver-node/blob/main/textDocument/src/main.ts#L440

The `index` is needed to sort as stable.

hrsh7th commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -255,45 +252,73 @@ function M.apply_text_edits(text_edits, bufnr)
+  end, text_edits)
+
+  -- Sort text_edits
+  table.sort(text_edits, function(a, b)
```
See https://github.com/microsoft/vscode-languageserver-node/blob/main/textDocument/src/main.ts#L355

hrsh7th commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -255,45 +252,73 @@ function M.apply_text_edits(text_edits, bufnr)
-    local A = {e.A[1] - start_line, e.A[2]}
-    local B = {e.B[1] - start_line, e.B[2]}
-    lines = M.set_lines(lines, A, B, e.lines)
+  -- Shrink overflowed ranges.
```
This is vim specialized implementation.

nvim_buf_set_text only accepts (row, col) tuple that is contained in buffer range.
However, Sometimes, Some LSP server sends  `+1` lines of range of TextEdit.

hrsh7th commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -1090,6 +1090,30 @@ describe('LSP', function()
         'å å ɧ 汉语 ↥ 🤦 🦄';
       }, buf_lines(1))
     end)
+    it('applies complex edits (reversed range)', function()
```
Add tests for reversed range.

mjlbach commented:
See also: https://github.com/neovim/neovim/pull/13703

Ping me when you are ready for review

hrsh7th commented:
@mjlbach I think it is ready for review.

hrsh7th commented:
I don't understand how to re-locate the cursor position if the cursor is contained in the edit range yet.

This PR doesn't re-locate the cursor in that case for now.

hrsh7th commented:
I didn't change the Lua code in the last force push. I'm just deleting the garbage comments.

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -255,45 +252,108 @@ function M.apply_text_edits(text_edits, bufnr)
+    }
+    vim.api.nvim_buf_set_text(bufnr, e.start_row, e.start_col, e.end_row, e.end_col, e.text)
+
+    local row_count = (e.end_row - e.start_row) + 1
+    if e.end_row < cursor.row then
+      cursor.row = cursor.row + (#e.text - row_count)
+      is_cursor_fixed = true
+    elseif e.end_row == cursor.row and e.end_col <= cursor.col then
+      cursor.row = cursor.row + (#e.text - row_count)
+      cursor.col = #e.text[#e.text] + (cursor.col - e.end_col)
+      if #e.text == 1 then
+        cursor.col = cursor.col + e.start_col
+      end
+      is_cursor_fixed = true
+    end
```
I think taking into account the cursor position here would deserve some comments that explain why this is necessary.

hrsh7th replied:
I've added the comments for it.

mjlbach commented:
@hrsh7th can you update the commit message? This PR is more than just fixing the cursor position, it should also note the other changes

hrsh7th committed:
feat(lsp): improve vim.lsp.util.apply_text_edits
- Fix the cursor position after applying TextEdits
- Support reversed range of TextEdit
- Invoke nvim_buf_set_text one by one

hrsh7th commented:
@mjlbach  I've updated the commit message.
